### ISSUE TITLE
Bundling haunted for a component library

### ISSUE THREAD

**Issue by d-Pixie** [matthewp/haunted#135]
We are trying to build a web components based component library and I continually get stuck on the issue where transpiled classes can't extend native classes.

As I see it there are two routes here:
* Bundle only our components and keep treating haunted as a third party library that will have to be imported in the consuming web page.
* Or bundle everything, including haunted in our bundle and not having to load any external libraries.

I have tried every combination I can think of for setting up Rollup for this and in the end I always run into one of two failure scenarios:

Either I bundle our code alone and treat haunted as an external dependancy. Then I run into `TypeError: Class constructor Hook cannot be invoked without 'new'` for a custom hook that I wrote. Right now the implementation of that hook is just a copy of the `use-state.js` from haunted in my node_modules, just to rule out that I did something wrong.

Or when I bundle haunted into our bundle I get basically the same error, but regarding HTMLElement instead, and the error message is a bit more verbose.

As far as I have been able to gather this has something to do with transpiled vs real ES6 classes, which would make sense except that for my simple bundle I don't do any transpiling, it is just a straight ES module and still throws.

**My question is**: If I want to build a bundle of haunted and my components to be loaded via a script tag in a browser, how do I setup the bundling?

My current Rollup config, the simple version:
```
import resolve from 'rollup-plugin-node-resolve'

const dependencies = Object.keys(require('./package.json').dependencies)

export default [
  {
    input: ['src/index.js'],
    output: {
      file: 'dist/index.js',
      format: 'es',
      sourcemap: true,
    },
    plugins: [
      resolve({
        customResolveOptions: {
          moduleDirectory: ['src'],
        },
      }),
    ],
    external: dependencies,
  }
];
```

My only dependency is haunted. I have also tried with various configurations and combinations of `rollup-plugin-commonjs` and `rollup-plugin-babel` and a ton of Babel 7 config to try to get it to leave the classes alone. But I can't get anything to build to a working version for the `iife` target, that I would need for this to execute in a browser as far as I understand.

Thankful for any help.

**Comment by chase-moskal** [comment]
@d-Pixie — hello friend

> We are trying to build a web components based component library

for this use-case, i'd recommend against bundling haunted with each component

my thinking is, from the perspective of the consumer of the components, when they import multiple component bundles from your library, you don't want each component bundle to contain its own copy of haunted

to avoid loading the same code multiple times, you need a way to resolve shared peer dependencies

generally, my strategy is to avoid any bundling for *library* code, and leave all of the resolution/bundling/minification and other such concerns up to the *consuming* application

- make all of your library code simple/lean/modern esnext modules
- the consuming project can then resolve peer dependencies and transpile/optimize the application for older browers at its own discretion — and there are several different tooling pathways ripe for the job
  - use npm/rollup/babel to bundle up the application
  - use [es-module-shims](https://github.com/guybedford/es-module-shims) as an importmap polyfill which allows the module/peer-dependency resolution to happen at runtime (this is a great option for small apps or during development)
    - i personally use my own experimental tool, [importly](https://github.com/chase-moskal/importly), to generate importmaps (to be used in conjunction with es-module-shims until importmaps are browser-native)
  - use [unpkg's ?module feature](https://unpkg.com/) to allow the unpkg server to resolve dependencies at runtime

i can't say i really understand the errors you're receiving, however, i suspect they might disappear if you take to heart the concepts i'm describing — do let us know more about these errors if they continue to be a problem for you

cheers,

:wave: Chase

**Comment by d-Pixie** [comment]
Hi @chase-moskal 

>  i'd recommend against bundling haunted with each component

We aren't bundling haunted with each component, but in the library itself. Right now the library is an all or nothing deal, you run the script and it registers all of the web components within it. Ideally most of them should be importable, that is only register when needed, but I haven't figured out a good way to do that yet :)

This is a library that we are going to use internally, so it doesn't have the same requirements that I would put on a public component library. Ideally I'd like to be able to build site specific bundles from it, containing only the components that a particular site needs, but for now I'm happy to just get it working at all.

The fact that it is internal is also a bit of a problem for some parts of your suggestions, I can't use unpkg for example, but I think that would be a possible way forward for the all in one approach.

I did consider this path before, loading haunted on the side so to speak, but landed with the bundle approach after reading up a bit about module support in browsers and how, for now, bundle loading affects page speed negatively du to how browsers have optimised javascript loading for bundles etc.

**Comment by chase-moskal** [comment]
> Right now the library is an all or nothing deal, you run the script and it registers all of the web components within it.

that's not a good restriction, there's no reason to load any component code that you don't need

> Ideally most of them should be importable, that is only register when needed, but I haven't figured out a good way to do that yet :)

let's talk about meeting that end

here's an example of a component library i've authored — take a look at [menutown/dist/register-all.js](https://unpkg.com/menutown@0.0.0/dist/register-all.js)

- for convenience, this registers all of the components in this library (there are only two in this example, `menu-system` and `menu-display`)
- however, there's no reason that the user would have to load both components if they only wanted one
- if the user only wanted `menu-display`, they could simply do this:
    ```js
    import {MenuDisplay} from "menutown/dist/components/menu-display.js"
    import {registerComponents} from "menutown/dist/toolbox/register-components.js"
    registerComponents({MenuDisplay})
    ```

in the above example, i can choose to load only one of the components that i need

and these components might use haunted, or maybe instead they'll use lit-element — it doesn't matter, different components can use different microframeworks, and there's no "all-or-nothing" library necessary

this is efficient, simple, and flexible

we can either use es-module-shims to resolve the dependencies at runtime via an importmap (perhaps using a utility like [importly](https://github.com/chase-moskal/importly) at build-time to generate the importmap) or perhaps more traditionally using npm and rollup in each application to produce a unique application bundle

> Ideally I'd like to be able to build site specific bundles from it, containing only the components that a particular site needs, but for now I'm happy to just get it working at all.

i fully understand the need to get things working as quickly as possible, however i'm certain that with a little more exploring this area, you can have your cake and eat it too

> I can't use unpkg

importly supports local serving from `node_modules` as an alternative to cdn's like unpkg or jsdelivr — as does traditional bundling systems like rollup

if you'd like, i'd be happy to chat more with you more about this on google hangouts or [discord](https://discord.gg/AXEJJP)

cheers mate,

:wave: Chase

**Comment by d-Pixie** [comment]
@chase-moskal I'll take you up on that :) I sent you an email.

I'll document our solution here for others to use when I have one.

**Comment by matthewp** [comment]
Figure this out?

**Comment by d-Pixie** [comment]
Not yet, no.

Do you have any recommendations on how to do this @matthewp? The combination of server side rendering, with Gatsby for example, and haunted doesn't seem easy to accomplish. At least not until lit-html adds propper SSR support. And bundling with babel doesn't seem to work, at least not if you write custom hooks ... Loading as a script tag with `type=module` does work. But support for import maps is a long way from being common and handling dependencies becomes a problem instead ...

That is without mentioning the complexity of wanting to have a large component library and just importing some components in a project. Normally webpack/rollup/parcel would solve that with tree shaking, but since I can't get that to build when using babel ...

I'll report back after talking to @chase-moskal, I was just about to email him about a meeting time 🙂

**Comment by d-Pixie** [comment]
So, what we decided to do in the end was a multi component repo using [Lerna](https://github.com/lerna/lerna). In large parts we modeled our setup after the [Google material design web components library](https://github.com/material-components/material-components-web-components). Our take can be found on Github as [oma-wc](https://github.com/oma-wc/oma).

The resulting setup allows us both to use individual components in WordPress or pure HTML/JS/CSS projects via `<script module src=""></script>` tags and bundle a bunch (but not all or nothing) of the modules together in React or Gatsby projects etc.

All in all this works ok for us. We still have some problems with lithtml not supporting server side rendering (an issue with Gatsby for example), and babel messing up the classes in some builds. So I'd love some pointers on that if you have them @matthewp or @chase-moskal 😄
